File: backend/app/schemas/appraisals.py
from __future__ import annotations
from typing import Optional, List
from datetime import datetime, date
from decimal import Decimal
from uuid import UUID
from pydantic import BaseModel, Field

from app.models.appraisals import (
    AppraisalPeriodType, AppraisalPeriodStatus, AppraisalStatus,
    FinalRating, DevelopmentPlanStatus, APAIAgentType, APAIRecStatus,
)


# ── Appraisal Period ──────────────────────────────────────────────────────────

class AppraisalPeriodCreate(BaseModel):
    period_code: str
    period_name: str
    period_type: AppraisalPeriodType
    start_date: date
    end_date: date
    review_start_date: Optional[date] = None
    review_end_date: Optional[date] = None
    notes: Optional[str] = None


class AppraisalPeriodUpdate(BaseModel):
    period_name: Optional[str] = None
    period_type: Optional[AppraisalPeriodType] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    review_start_date: Optional[date] = None
    review_end_date: Optional[date] = None
    status: Optional[AppraisalPeriodStatus] = None
    notes: Optional[str] = None


class AppraisalPeriodOut(BaseModel):
    appraisal_period_id: UUID
    period_code: str
    period_name: str
    period_type: AppraisalPeriodType
    start_date: date
    end_date: date
    review_start_date: Optional[date]
    review_end_date: Optional[date]
    status: AppraisalPeriodStatus
    notes: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# ── Appraisal Template ────────────────────────────────────────────────────────

class AppraisalTemplateCreate(BaseModel):
    template_code: str
    template_name: str
    department_id: Optional[str] = None
    role_id: Optional[str] = None
    employee_grade_id: Optional[str] = None
    kpi_weight_pct: Decimal = Field(default=Decimal("60.00"))
    competency_weight_pct: Decimal = Field(default=Decimal("40.00"))
    active_flag: bool = True
    notes: Optional[str] = None


class AppraisalTemplateOut(BaseModel):
    template_id: UUID
    template_code: str
    template_name: str
    department_id: Optional[str]
    role_id: Optional[str]
    employee_grade_id: Optional[str]
    kpi_weight_pct: Decimal
    competency_weight_pct: Decimal
    active_flag: bool
    notes: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# ── KPI Line ──────────────────────────────────────────────────────────────────

class KPILineCreate(BaseModel):
    kpi_name: str
    kpi_description: Optional[str] = None
    target_value: Optional[Decimal] = None
    actual_value: Optional[Decimal] = None
    unit: Optional[str] = None
    weight_pct: Decimal = Field(default=Decimal("0.00"))
    self_score: Optional[Decimal] = None
    comments: Optional[str] = None


class KPILineUpdate(BaseModel):
    actual_value: Optional[Decimal] = None
    self_score: Optional[Decimal] = None
    manager_score: Optional[Decimal] = None
    final_score: Optional[Decimal] = None
    comments: Optional[str] = None


class KPILineOut(BaseModel):
    kpi_line_id: UUID
    appraisal_id: UUID
    kpi_name: str
    kpi_description: Optional[str]
    target_value: Optional[Decimal]
    actual_value: Optional[Decimal]
    unit: Optional[str]
    weight_pct: Decimal
    self_score: Optional[Decimal]
    manager_score: Optional[Decimal]
    final_score: Optional[Decimal]
    comments: Optional[str]

    class Config:
        from_attributes = True


# ── Competency Line ───────────────────────────────────────────────────────────

class CompetencyLineCreate(BaseModel):
    competency_name: str
    description: Optional[str] = None
    weight_pct: Decimal = Field(default=Decimal("0.00"))
    self_rating: Optional[Decimal] = None
    comments: Optional[str] = None


class CompetencyLineUpdate(BaseModel):
    self_rating: Optional[Decimal] = None
    manager_rating: Optional[Decimal] = None
    final_rating: Optional[Decimal] = None
    comments: Optional[str] = None


class CompetencyLineOut(BaseModel):
    competency_line_id: UUID
    appraisal_id: UUID
    competency_name: str
    description: Optional[str]
    weight_pct: Decimal
    self_rating: Optional[Decimal]
    manager_rating: Optional[Decimal]
    final_rating: Optional[Decimal]
    comments: Optional[str]

    class Config:
        from_attributes = True


# ── Development Plan ──────────────────────────────────────────────────────────

class DevelopmentPlanCreate(BaseModel):
    goal_title: str
    action_plan: Optional[str] = None
    due_date: Optional[date] = None
    owner_id: Optional[str] = None
    owner_name: Optional[str] = None
    notes: Optional[str] = None


class DevelopmentPlanUpdate(BaseModel):
    goal_title: Optional[str] = None
    action_plan: Optional[str] = None
    due_date: Optional[date] = None
    status: Optional[DevelopmentPlanStatus] = None
    notes: Optional[str] = None


class DevelopmentPlanOut(BaseModel):
    development_plan_id: UUID
    appraisal_id: UUID
    employee_id: str
    goal_title: str
    action_plan: Optional[str]
    due_date: Optional[date]
    owner_id: Optional[str]
    owner_name: Optional[str]
    status: DevelopmentPlanStatus
    notes: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# ── Appraisal Record ──────────────────────────────────────────────────────────

class AppraisalRecordCreate(BaseModel):
    appraisal_period_id: UUID
    employee_id: str
    employee_name: Optional[str] = None
    manager_id: Optional[str] = None
    manager_name: Optional[str] = None
    department_id: Optional[str] = None
    department_name: Optional[str] = None
    template_id: Optional[UUID] = None
    notes: Optional[str] = None


class SelfSubmitRequest(BaseModel):
    self_score: Optional[Decimal] = None
    notes: Optional[str] = None


class ManagerReviewRequest(BaseModel):
    manager_score: Optional[Decimal] = None
    increment_recommendation: Optional[Decimal] = None
    promotion_recommendation: bool = False
    notes: Optional[str] = None


class HRReviewRequest(BaseModel):
    final_score: Optional[Decimal] = None
    final_rating: Optional[FinalRating] = None
    hr_notes: Optional[str] = None
    calibration_notes: Optional[str] = None


class FinalizeRequest(BaseModel):
    final_score: Optional[Decimal] = None
    final_rating: Optional[FinalRating] = None
    increment_recommendation: Optional[Decimal] = None
    promotion_recommendation: Optional[bool] = None
    hr_notes: Optional[str] = None


class AppraisalRecordOut(BaseModel):
    appraisal_id: UUID
    appraisal_period_id: UUID
    employee_id: str
    employee_name: Optional[str]
    manager_id: Optional[str]
    manager_name: Optional[str]
    department_id: Optional[str]
    department_name: Optional[str]
    template_id: Optional[UUID]
    status: AppraisalStatus
    self_score: Optional[Decimal]
    manager_score: Optional[Decimal]
    final_score: Optional[Decimal]
    final_rating: Optional[FinalRating]
    increment_recommendation: Optional[Decimal]
    promotion_recommendation: bool
    self_submitted_at: Optional[datetime]
    manager_reviewed_at: Optional[datetime]
    hr_reviewed_at: Optional[datetime]
    finalized_at: Optional[datetime]
    hr_notes: Optional[str]
    calibration_notes: Optional[str]
    notes: Optional[str]
    created_at: datetime
    kpi_lines: List[KPILineOut] = []
    competency_lines: List[CompetencyLineOut] = []
    development_plans: List[DevelopmentPlanOut] = []

    class Config:
        from_attributes = True


# ── Dashboard ─────────────────────────────────────────────────────────────────

class AppraisalDashboard(BaseModel):
    total_records: int
    self_review_pending: int
    manager_review_pending: int
    hr_review_pending: int
    calibration_pending: int
    completed: int
    avg_final_score: Optional[float]
    rating_distribution: dict
    department_avg_scores: List[dict]
    pending_development_plans: int
    promotion_recommendations: int


# ── AI ────────────────────────────────────────────────────────────────────────

class APAIRecOut(BaseModel):
    rec_id: UUID
    agent_type: APAIAgentType
    status: APAIRecStatus
    title: str
    body: str
    appraisal_id: Optional[UUID]
    employee_id: Optional[str]
    department_id: Optional[str]
    score: Optional[int]
    actioned_by: Optional[str]
    actioned_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class APAIRecAck(BaseModel):
    status: APAIRecStatus
    actioned_by: Optional[str] = None


AppraisalRecordOut.model_rebuild()

File: frontend/src/lib/appraisals.ts
const BASE = `${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/api/v1/appraisals`;

async function req<T>(path: string, opts?: RequestInit): Promise<T> {
  const r = await fetch(`${BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...opts,
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error(err.detail ?? r.statusText);
  }
  if (r.status === 204) return undefined as T;
  return r.json();
}

// ── Types ─────────────────────────────────────────────────────────────────────

export type AppraisalPeriodType = "annual" | "semi_annual" | "quarterly" | "probation" | "project_based" | "custom";
export type AppraisalPeriodStatus = "draft" | "active" | "closed" | "archived";
export type AppraisalStatus = "draft" | "self_review" | "manager_review" | "hr_review" | "calibration" | "completed" | "rejected";
export type FinalRating = "excellent" | "good" | "meets_expectations" | "improvement_needed";
export type DevelopmentPlanStatus = "open" | "in_progress" | "completed" | "cancelled";
export type APAIAgentType = "performance_insight" | "calibration_risk" | "development_plan";
export type APAIRecStatus = "pending" | "acknowledged" | "actioned" | "dismissed";

export interface AppraisalPeriod {
  appraisal_period_id: string;
  period_code: string;
  period_name: string;
  period_type: AppraisalPeriodType;
  start_date: string;
  end_date: string;
  review_start_date: string | null;
  review_end_date: string | null;
  status: AppraisalPeriodStatus;
  notes: string | null;
  created_at: string;
}

export interface AppraisalTemplate {
  template_id: string;
  template_code: string;
  template_name: string;
  department_id: string | null;
  role_id: string | null;
  employee_grade_id: string | null;
  kpi_weight_pct: number;
  competency_weight_pct: number;
  active_flag: boolean;
  notes: string | null;
  created_at: string;
}

export interface KPILine {
  kpi_line_id: string;
  appraisal_id: string;
  kpi_name: string;
  kpi_description: string | null;
  target_value: number | null;
  actual_value: number | null;
  unit: string | null;
  weight_pct: number;
  self_score: number | null;
  manager_score: number | null;
  final_score: number | null;
  comments: string | null;
}

export interface CompetencyLine {
  competency_line_id: string;
  appraisal_id: string;
  competency_name: string;
  description: string | null;
  weight_pct: number;
  self_rating: number | null;
  manager_rating: number | null;
  final_rating: number | null;
  comments: string | null;
}

export interface DevelopmentPlan {
  development_plan_id: string;
  appraisal_id: string;
  employee_id: string;
  goal_title: string;
  action_plan: string | null;
  due_date: string | null;
  owner_id: string | null;
  owner_name: string | null;
  status: DevelopmentPlanStatus;
  notes: string | null;
  created_at: string;
}

export interface AppraisalRecord {
  appraisal_id: string;
  appraisal_period_id: string;
  employee_id: string;
  employee_name: string | null;
  manager_id: string | null;
  manager_name: string | null;
  department_id: string | null;
  department_name: string | null;
  template_id: string | null;
  status: AppraisalStatus;
  self_score: number | null;
  manager_score: number | null;
  final_score: number | null;
  final_rating: FinalRating | null;
  increment_recommendation: number | null;
  promotion_recommendation: boolean;
  self_submitted_at: string | null;
  manager_reviewed_at: string | null;
  hr_reviewed_at: string | null;
  finalized_at: string | null;
  hr_notes: string | null;
  calibration_notes: string | null;
  notes: string | null;
  created_at: string;
  kpi_lines: KPILine[];
  competency_lines: CompetencyLine[];
  development_plans: DevelopmentPlan[];
}

export interface AppraisalDashboard {
  total_records: number;
  self_review_pending: number;
  manager_review_pending: number;
  hr_review_pending: number;
  calibration_pending: number;
  completed: number;
  avg_final_score: number | null;
  rating_distribution: Record<string, number>;
  department_avg_scores: { department: string; avg_score: number }[];
  pending_development_plans: number;
  promotion_recommendations: number;
}

export interface APAIRec {
  rec_id: string;
  agent_type: APAIAgentType;
  status: APAIRecStatus;
  title: string;
  body: string;
  appraisal_id: string | null;
  employee_id: string | null;
  department_id: string | null;
  score: number | null;
  actioned_by: string | null;
  actioned_at: string | null;
  created_at: string;
}

// ── Color / Label Maps ────────────────────────────────────────────────────────

export const PERIOD_STATUS_COLOR: Record<AppraisalPeriodStatus, string> = {
  draft: "bg-gray-100 text-gray-700",
  active: "bg-green-100 text-green-700",
  closed: "bg-blue-100 text-blue-700",
  archived: "bg-gray-200 text-gray-500",
};

export const APPRAISAL_STATUS_COLOR: Record<AppraisalStatus, string> = {
  draft: "bg-gray-100 text-gray-700",
  self_review: "bg-yellow-100 text-yellow-700",
  manager_review: "bg-orange-100 text-orange-700",
  hr_review: "bg-purple-100 text-purple-700",
  calibration: "bg-indigo-100 text-indigo-700",
  completed: "bg-green-100 text-green-700",
  rejected: "bg-red-100 text-red-700",
};

export const APPRAISAL_STATUS_LABEL: Record<AppraisalStatus, string> = {
  draft: "Draft",
  self_review: "Self Review",
  manager_review: "Manager Review",
  hr_review: "HR Review",
  calibration: "Calibration",
  completed: "Completed",
  rejected: "Rejected",
};

export const RATING_COLOR: Record<FinalRating, string> = {
  excellent: "bg-green-100 text-green-800",
  good: "bg-blue-100 text-blue-800",
  meets_expectations: "bg-yellow-100 text-yellow-800",
  improvement_needed: "bg-red-100 text-red-800",
};

export const RATING_LABEL: Record<FinalRating, string> = {
  excellent: "Excellent (90–100)",
  good: "Good (75–89)",
  meets_expectations: "Meets Expectations (60–74)",
  improvement_needed: "Needs Improvement (<60)",
};

export const PLAN_STATUS_COLOR: Record<DevelopmentPlanStatus, string> = {
  open: "bg-blue-100 text-blue-700",
  in_progress: "bg-yellow-100 text-yellow-700",
  completed: "bg-green-100 text-green-700",
  cancelled: "bg-gray-100 text-gray-500",
};

export const PERIOD_TYPE_LABEL: Record<AppraisalPeriodType, string> = {
  annual: "Annual",
  semi_annual: "Semi-Annual",
  quarterly: "Quarterly",
  probation: "Probation",
  project_based: "Project-Based",
  custom: "Custom",
};

// ── API Client ────────────────────────────────────────────────────────────────

export const appraisalsApi = {
  // Dashboard
  getDashboard: (period_id?: string) =>
    req<AppraisalDashboard>(`/dashboard${period_id ? `?period_id=${period_id}` : ""}`),

  // Periods
  createPeriod: (d: object) => req<AppraisalPeriod>("/periods", { method: "POST", body: JSON.stringify(d) }),
  listPeriods: (status?: string) => req<AppraisalPeriod[]>(`/periods${status ? `?status=${status}` : ""}`),
  getPeriod: (id: string) => req<AppraisalPeriod>(`/periods/${id}`),
  updatePeriod: (id: string, d: object) => req<AppraisalPeriod>(`/periods/${id}`, { method: "PATCH", body: JSON.stringify(d) }),

  // Templates
  createTemplate: (d: object) => req<AppraisalTemplate>("/templates", { method: "POST", body: JSON.stringify(d) }),
  listTemplates: (active?: boolean) => req<AppraisalTemplate[]>(`/templates${active ? "?active_only=true" : ""}`),
  getTemplate: (id: string) => req<AppraisalTemplate>(`/templates/${id}`),

  // Records
  createRecord: (d: object) => req<AppraisalRecord>("/", { method: "POST", body: JSON.stringify(d) }),
  listRecords: (params?: { period_id?: string; employee_id?: string; department_id?: string; status?: string }) => {
    const q = new URLSearchParams();
    if (params?.period_id) q.set("period_id", params.period_id);
    if (params?.employee_id) q.set("employee_id", params.employee_id);
    if (params?.department_id) q.set("department_id", params.department_id);
    if (params?.status) q.set("status", params.status);
    return req<AppraisalRecord[]>(`/?${q.toString()}`);
  },
  getRecord: (id: string) => req<AppraisalRecord>(`/${id}`),
  selfSubmit: (id: string, d: object) => req<AppraisalRecord>(`/${id}/self-submit`, { method: "POST", body: JSON.stringify(d) }),
  managerReview: (id: string, d: object) => req<AppraisalRecord>(`/${id}/manager-review`, { method: "POST", body: JSON.stringify(d) }),
  hrReview: (id: string, d: object) => req<AppraisalRecord>(`/${id}/hr-review`, { method: "POST", body: JSON.stringify(d) }),
  finalize: (id: string, d: object) => req<AppraisalRecord>(`/${id}/finalize`, { method: "POST", body: JSON.stringify(d) }),
  reject: (id: string, notes?: string) =>
    req<AppraisalRecord>(`/${id}/reject${notes ? `?notes=${encodeURIComponent(notes)}` : ""}`, { method: "POST" }),
  calculateScores: (id: string) => req<AppraisalRecord>(`/${id}/calculate-scores`, { method: "POST" }),

  // KPI Lines
  addKPILine: (appraisalId: string, d: object) =>
    req<KPILine>(`/${appraisalId}/kpi-lines`, { method: "POST", body: JSON.stringify(d) }),
  updateKPILine: (id: string, d: object) =>
    req<KPILine>(`/kpi-lines/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  deleteKPILine: (id: string) => req<void>(`/kpi-lines/${id}`, { method: "DELETE" }),

  // Competency Lines
  addCompetencyLine: (appraisalId: string, d: object) =>
    req<CompetencyLine>(`/${appraisalId}/competency-lines`, { method: "POST", body: JSON.stringify(d) }),
  updateCompetencyLine: (id: string, d: object) =>
    req<CompetencyLine>(`/competency-lines/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  deleteCompetencyLine: (id: string) => req<void>(`/competency-lines/${id}`, { method: "DELETE" }),

  // Development Plans
  addDevelopmentPlan: (appraisalId: string, employeeId: string, d: object) =>
    req<DevelopmentPlan>(`/${appraisalId}/development-plans?employee_id=${employeeId}`, { method: "POST", body: JSON.stringify(d) }),
  updateDevelopmentPlan: (id: string, d: object) =>
    req<DevelopmentPlan>(`/development-plans/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  listDevelopmentPlans: (params?: { employee_id?: string; status?: string }) => {
    const q = new URLSearchParams();
    if (params?.employee_id) q.set("employee_id", params.employee_id);
    if (params?.status) q.set("status", params.status);
    return req<DevelopmentPlan[]>(`/development-plans/list?${q.toString()}`);
  },

  // Reports
  reportCompletion: (period_id: string) => req<object>(`/reports/completion?period_id=${period_id}`),
  reportRatingDist: (period_id?: string) =>
    req<object[]>(`/reports/rating-distribution${period_id ? `?period_id=${period_id}` : ""}`),
  reportPromotions: (period_id?: string) =>
    req<object[]>(`/reports/promotions${period_id ? `?period_id=${period_id}` : ""}`),

  // AI
  runPerformanceInsight: () => req<APAIRec[]>("/ai/run-performance-insight", { method: "POST" }),
  runCalibrationRisk: () => req<APAIRec[]>("/ai/run-calibration-risk", { method: "POST" }),
  runDevelopmentPlan: () => req<APAIRec[]>("/ai/run-development-plan", { method: "POST" }),
  listAIRecs: (status?: string) => req<APAIRec[]>(`/ai/recommendations${status ? `?status=${status}` : ""}`),
  ackAIRec: (id: string, d: object) =>
    req<APAIRec>(`/ai/recommendations/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
};
